### PR TITLE
[Android, iOS] Fix CarouselView Issues

### PR DESCRIPTION
**Pull request by jsuarezruiz**
### Description of Change ###

Changes in **CarouselView** to fix some issues.

### Issues Resolved ### 
- Fixed the calculations of element **size** on Android.
- Fixed calculations of **Insets (PeekAreaInsets)** on Android.
- Protected NRE setting the **CurrentItem**.
- Nothing happened when changing the **ItemsSource** on Android, it has been fixed.
- Fixed **ItemSpacing** on Android.
- The **NumberOfSideItems** property now works on Android.
- Improved **IsSwipeEnabled** property. Now, disabling the swipe can continue touching the child items (fire child events). 
- Fixed crash on Android sample.

You can see a summary of the status in: https://gist.github.com/jsuarezruiz/9c071593dc05f4e91c9f93f1f8f8d4ab

### API Changes ###
 
 None

### Platforms Affected ### 

- iOS
- Android

### Behavioral/Visual Changes ###

None

#### Before

_**NOTE:** Screenshots from master branch._

Nothing happens changing the ItemsSource:
itemssource-before
ItemSpacing not working:
itemspacing-before
NumberOfSideItems not working:
Captura de pantalla 2019-09-03 a las 13 53 03
PeekAreaInsets not working:
Captura de pantalla 2019-09-03 a las 13 53 17
Wrong item size:
Captura de pantalla 2019-09-03 a las 13 49 42

#### After
Fixed changing ItemsSource (Android & iOS):
itemssource-after
Fixed ItemSpacing:
itemspacing-after2
Center items after changes ItemSpacing:
update-itemspacing
Fixed NumberOfSideItems:
numberofsideitems-after
Fixed PeekAreaInsets:
peekareainsets-after
Fixed size:
size-after
Improved IsSwipeEnabled property. Now, disabling the swipe can continue touching the child items (fire child events). 
isswipeenable-droid-fix
Fixed samples:
carouselviewxaml

### Testing Procedure ###
Open Core Gallery and select the CarouselView samples. Test all the main properties.

### PR Checklist ###

- [ ] Has automated tests 
- [ ] Rebased on top of the target branch at time of PR
- [ ] Changes adhere to coding standard

### CODE REVIEW

File: Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs [hunk 1]
```diff
@@ -18,15 +19,21 @@ public class CarouselViewRenderer : ItemsViewRenderer<ItemsView, ItemsViewAdapte
...
 			{
+				if(_itemDecoration != null)
+				{
+					_itemDecoration.Dispose();
+					_itemDecoration = null;
+				}
 
-			}
+                _layout = null;
+            }
```
Comment (rmarinho) [inline Xamarin.Forms.Platform.Android/CollectionView/CarouselViewRenderer.cs]: Fix formatting

**Comment by rmarinho** [comment]
Can we target this to 4.3.0 ?

File: GitInfo.txt [hunk 2]
```diff
@@ -1 +1 @@
-4.3.0
+4.4.0
```
Comment (jfversluis) [inline GitInfo.txt]: Since the retarget to 4.3 I don't think this is correct?

Comment (jfversluis) [inline GitInfo.txt]:
```suggestion
4.3.0
```

File: Xamarin.Forms.Controls/CoreGalleryPages/CarouselViewCoreGalleryPage.cs [hunk 3]
```diff
@@ -8,6 +8,8 @@ namespace Xamarin.Forms.Controls
 	[Preserve(AllMembers = true)]
 	internal class CarouselViewCoreGalleryPage : CoreGalleryPage<CarouselView>
 	{
+		private object _currentItem;
```
Comment (jfversluis) [inline Xamarin.Forms.Controls/CoreGalleryPages/CarouselViewCoreGalleryPage.cs]:
```suggestion
		object _currentItem;
```

File: Xamarin.Forms.Core/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -79,6 +79,7 @@
 [assembly: StyleProperty("vertical-align", typeof(ITextAlignmentElement), nameof(TextAlignmentElement.VerticalTextAlignmentProperty))]
 [assembly: StyleProperty("visibility", typeof(VisualElement), nameof(VisualElement.IsVisibleProperty), Inherited = true)]
 [assembly: StyleProperty("width", typeof(VisualElement), nameof(VisualElement.WidthRequestProperty))]
+[assembly: StyleProperty("letter-spacing", typeof(ITextElement), nameof(TextElement.CharacterSpacingProperty), Inherited = true)]
```
Comment (jfversluis) [inline Xamarin.Forms.Core/Properties/AssemblyInfo.cs]: Is this part of this change?

File: Xamarin.Forms.Xaml.UnitTests/LabelHtml.xaml [hunk 5]
```diff
@@ -0,0 +1,34 @@
+﻿<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jfversluis) [inline Xamarin.Forms.Xaml.UnitTests/LabelHtml.xaml]: This (and below) are definitely not part of this change 😄 do we need a rebase?

**Comment by vniehues** [comment]
Since this change i get a "cannot-access-a-disposed-object exception" when adding an item to my CarouselView ItemsSource (ObservableCollection).

**Comment by jfversluis** [comment]
@vniehues sorry about that! Could you please see if there is already an issue opened for that and if not open one? Please add all relevant info and a reproduction project. 

Thanks!

**Comment by vniehues** [comment]
Yeah i'll do a sanity check and build a repro project to see if its really "your fault" or if i srewed something up. 
I'll let you about it.

**Comment by vniehues** [comment]
Okay so i cant reproduce it which might mean that there is an error on my side.

However it seems that there was a related bug with the ListView in the past, where items where not recycled correctly: [#2406](https://github.com/xamarin/Xamarin.Forms/issues/2406).

From my error message:

_**System.ObjectDisposedException:** 'Cannot access a disposed object.
Object name: 'Xamarin.Forms.Platform.Android.ItemsViewAdapter`2[[Xamarin.Forms.ItemsView, Xamarin.Forms.Core, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null],[Xamarin.Forms.Platform.Android.IItemsViewSource, Xamarin.Forms.Platform.Android, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null]]'.'_

It looks like the ItemsViewAdapter does dispose the items to early. 

Side-Note: it does currently not matter if i try to insert the new item at the top (Items[0]) or at the End (Items.Add()). 

I hope that these information somehow help you narrowing it down.
